sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/aggregator/ExponentialHistogramIndexer.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.metrics.internal.aggregator;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+class ExponentialHistogramIndexer {
+
+  private static final Map<Integer, ExponentialHistogramIndexer> cache = new ConcurrentHashMap<>();
+
+  /** Bit mask used to isolate exponent of IEEE 754 double precision number. */
+  private static final long EXPONENT_BIT_MASK = 0x7FF0000000000000L;
+
+  /** Bit mask used to isolate the significand of IEEE 754 double precision number. */
+  private static final long SIGNIFICAND_BIT_MASK = 0xFFFFFFFFFFFFFL;
+
+  /** Bias used in representing the exponent of IEEE 554 double precision number. */
+  private static final int EXPONENT_BIAS = 1023;
+
+  /**
+   * The number of bits used to represent the significand of IEEE 754 double precision number,
+   * excluding the implicit bit.
+   */
+  private static final int SIGNIFICAND_WIDTH = 52;
+
+  private static final double LOG_BASE2_E = 1D / Math.log(2);
+
+  private final int scale;
+  private final double scaleFactor;
+
+  private ExponentialHistogramIndexer(int scale) {
+    this.scale = scale;
+    this.scaleFactor = computeScaleFactor(scale);
+  }
+
+  /** Get an indexer for the given scale. Indexers are cached and reused for */
+  public static ExponentialHistogramIndexer get(int scale) {
+    return cache.computeIfAbsent(scale, unused -> new ExponentialHistogramIndexer(scale));
+  }
+
+  /**
+   * Compute the index for the given value.
+   *
+   * <p>The algorithm to retrieve the index is specified in the <a
+   * href="https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/datamodel.md#exponential-buckets">OpenTelemetry
+   * specification</a>.
+   *
+   * @param value Measured value (must be non-zero).
+   * @return the index of the bucket which the value maps to.
+   */
+  int computeIndex(double value) {
+    double absValue = Math.abs(value);
+    if (scale > 0) {
+      return getIndexByLogarithm(absValue);
+    }
+    if (scale == 0) {
+      return mapToIndexScaleZero(absValue);
+    }
+    // scale < 0
+    return mapToIndexScaleZero(absValue) >> -scale;
+  }
```

Comment:
I think we can reduce the modality of the code here and use a single algorithm to compute the index for all values of scale:
 
If we apply the log property of `log(x^n) = n * log(x)` we can get rid of the `getIndexByLogarithm`:
```java
escaleFactor = Math.scalb(LOG_BASE2_E, scale);
return (int) Math.ceil(Math.log(value) * scaleFactor) - 1;

// this is equivalent to

escaleFactor = Math.scalb(1, scale);
return mapToIndexScaleZero(Math.pow(value, scaleFactor));
```

Hence:

```java
 int computeIndex(double value) {
    double absValue = Math.abs(value);
    return mapToIndexScaleZero(Math.pow(absValue, scaleFactor));
  }

  private static double computeScaleFactor(int scale) {
    return Math.scalb(1, scale);
  }
```

The only caveat is that we are not going to round the subnormal values anymore. But that is ok because [the spec allows it](https://opentelemetry.io/docs/reference/specification/metrics/datamodel/#scale-zero-extract-the-exponent). We would need to update the test cases of `valueToIndex_ScaleNegFour` for the section `// Min and subnormal values`

On a side note, I think we should change the method name `mapToIndexScaleZero` to something like `getIeee754Exponent` or something like that.

Reply:
Took me a while to figure out why this worked but I eventually got there! Brought it up to @jmacd who correctly points out that this code will overflow. Specifically, anytime we're trying to determine the index of large values (i.e. values near `Double.MAX_VALUE`) and we have a positive scale factor, we'll be calling `Math.pow(value, Math.scalb(1, scale))`.

Reply:
It was worth trying :). 

Why we are not using the log based calculation everywhere then? What is the reason for trying different implementations based on the scale value? Would it be performance or a different reason?

One reason might be that extracting the ieeeExponent might produce a more exact mapping to index. But [the spec says that it is ok if they are not exact](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/data-model.md#producer-expectations). However if that is the case, is it ok to have two different behaviors depending on the scale (exact vs non-exact)?

Reply:
Implementations are recommended to be exact for powers-of-two. I think that using a logarithm implementation exclusively is fine, but I recommend testing near the +Inf and 0x1p-1022 boundaries. The exponent-based implementation is both faster and more exact than the logarithm-based implementation. If you were going to use a table-lookup-based approach for the positive scales (which I don't recommend, but we've seen several prototypes of), then you would might want the exponent-based implementation for scales <= 0.

Reply:
I think I'm ok with the current approach. It is consistent across the same scale so I think that is fine.

Is it worth adding some comments in the code regarding the reason why we are mixing the approaches? (i.e.: exponent-based is faster and exact, but can only be used for 0 or negative scales...) or is it referencing the spec enough?

Reply:
I've added some comments which help explain the mixed approach, but which still link to the spec for exact details of the algorithm. 

I'm going to mark this comment as resolved. Please unresolve it if you think there are still points to discuss 🙂